### PR TITLE
Fix completion handler replacement warnings

### PR DESCRIPTION
**Pull request by etcwilde**
This patch fixes a couple bugs in the completion-handler async functionality.
First, it fixes the errant diagnostic in `defer` and synchronous functions that are nested inside of an asynchronous context. Second, it fixes the missing diagnostics/crashing when diagnosing top-level code.

Since `@CompiletionHandlerAsync` ended up being an availability check, I've wrapped it into the asynchronous usage part of the availability diagnostics. This of course means that we can't use `-disable-availability-checking` and get these diagnostics, but it fits better in the model and doesn't require a separate pass.

rdar://85614267
rdar://86183022

### CODE REVIEW

**Comment by etcwilde** [comment]
@swift-ci please test

File: lib/Sema/TypeCheckAvailability.cpp [hunk 1]
```diff
@@ -3573,12 +3573,11 @@ bool ExprAvailabilityWalker::diagnoseDeclRefAvailability(
   return false;
 }
 
-/// Diagnose uses of API annotated '@unavailableFromAsync' when used from
-/// asynchronous contexts.
-/// Returns true if a diagnostic was emitted, false otherwise.
+/// Diagnose misuses of asynchronous API in asynchronous contexts.
```
Comment (bnbarham) [inline lib/Sema/TypeCheckAvailability.cpp:3576]: This is diagnosing both now right 🤔? Ie. use of a sync API in async context when there's an async alternative as well as use of an unavailable async API?

**Review by bnbarham — APPROVED** [review]

File: lib/Sema/TypeCheckAvailability.cpp (hunk 1 above)
Comment (etcwilde) [reply lib/Sema/TypeCheckAvailability.cpp:3576]: Yep

Comment (bnbarham) [reply lib/Sema/TypeCheckAvailability.cpp:3576]: I'm not sure that's clear to me from `asynchronous API in asynchronous context`, though I suppose you could argue that you're misusing the asynchronous API by using the synchronous API 🤷

Comment (etcwilde) [reply lib/Sema/TypeCheckAvailability.cpp:3576]: Something like that. I can change the comment if that's what you're asking for.

Comment (bnbarham) [reply lib/Sema/TypeCheckAvailability.cpp:3576]: Sure, could even just drop the first asynchronous really. Or expand it. Up to you.

Comment (etcwilde) [reply lib/Sema/TypeCheckAvailability.cpp]: Done

**Comment by etcwilde** [comment]
@swift-ci please test

**Comment by etcwilde** [comment]
@swift-ci please smoke test

**Comment by etcwilde** [comment]
For some reason, my computer had no issues with the missing availability checks. ¯\_(ツ)_/¯ Anyway, I think that should do it.

**Comment by etcwilde** [comment]
@swift-ci  please smoke test

**Comment by etcwilde** [comment]
@swift-ci please smoke test

**Review by hamishknight — APPROVED** [review]

**Pull request merged**